### ISSUE TITLE
Reuse tensordot memory tricks in matmul

### ISSUE THREAD

**Issue by mrocklin** [dask/dask#6874]
Currently the dask array matmul implemenation is similar to the tensordot implemenation (understandably) but lacks a clever trick that tensordot uses to avoid contractions in blockwise (which can result in large pileups of intermediate results).  

https://github.com/dask/dask/blob/ff3ea6c74e8736a5823d648defcf9164f083d266/dask/array/routines.py#L290-L329

Tensordot uses only outer products in blockwise (there are no shared indices in the blockwise call) , and then uses a `sum` call to reduce these.  Because sum uses tree reductions, this results in nicer memory usage.  Matmul doesn't use this trick.

I briefly tried replacing the blockwise call in `matmul` with a call to our internal `dot` function, but various tests failed.  It would be good to do one of two things:

1.  Find a way to back matmul by tensordot, so that the reduction trick is shared with a common codebase
2.  Replicate the reduction trick in matmul

This was originally reported by @eric-czech and @tomwhite in https://github.com/pystatgen/sgkit/issues/375#issuecomment-731060672

**Comment by tomwhite** [comment]
Unlike `matmul`, `tensordot` doesn't broadcast leading axes, so I'm not sure it's possible to rewrite `matmul` in terms of `tensordot`.

For example:
```python
>>> x = np.random.random((7, 11, 15))
>>> y = np.random.random((7, 15, 19))
>>> (x @ y).shape
(7, 11, 19)
>>> np.tensordot(x, y, ((2,), (1,))).shape
(7, 11, 7, 19)
```

It is possible to implement `matmul` using `einsum` though, see https://github.com/dask/dask/compare/9472b69c3333c3767fe9a500ea6c310d83cfb4c6...tomwhite:matmul-memory-usage, so that might be a way to solve this issue, since Dask's `einsum` implementation doesn't use `concatenate=True`.

**Comment by mrocklin** [comment]
If that works, then great :)

**Comment by eric-czech** [comment]
FWIW, I tried many different combinations of inputs (always two arrays of different shapes/chunkings) and found that there appear to no large differences between `matmul` and `einsum` as @tomwhite currently has it parameterized in that diff: https://gist.github.com/eric-czech/3c39785ebb08088775326154bbeddcc2

The conclusions I draw from this are:

1. Neither supports scalars
2. The `einsum` version produces results that it perhaps shouldn't for `matmul`, e.g. multiplying a `(10,1)` array by another `(10,1)` is possible w/ `einsum` but not the current `matmul` function.  Similarly, multiplying a `(1,1)` array by a `(10,1)` is ok w/ `einsum`.  There are many more cases like that but they seem to generalize as `einsum` supporting repetition of values in contracted axes of unit length, which `matmul` doesn't currently do.

**Comment by mrocklin** [comment]
I suspect that there will be some other challenges:

   1. Broadcasting rules for matmul may be atypical as the number of
   dimensions increases beyond 2.  I don't have a strong memory here but
   recall it being strange in some way
   2. Some numpy-like libraries (sparse, cupy) probably implement
   np.matmul, but not einsum.  It's probably better on our part if we use the
   np.matmul call on a per-chunk basis when we can

**Comment by tomwhite** [comment]
> I tried many different combinations of inputs

Thanks for trying those out @eric-czech. For the cases that differ, we could check that the axis lengths are compatible and raise an error like `matmul` does (if we decided to go down the `einsum` route).

> It's probably better on our part if we use the np.matmul call on a per-chunk basis when we can

`matmul` will always contract, so I don't see how we can use the `tensordot` trick with it (which relies on *not* contracting in the chunks, then doing the contraction via reductions). Using `tensordot` on the chunks is a possibility for some cases though (basically if either array has 2 dimensions or less).

**Comment by mrocklin** [comment]
The tensordot trick actually *does* contract in the chunks (otherwise the intermediate results would be possibly huge).  

However, we lie to the intermediate Dask Array so that it thinks that it hasn't contracted and that it should do the sum appropriately.   The intermediate dask array has shape and chunk structure metadata that is intentionally incorrect.

**Comment by tomwhite** [comment]
Ah ok - it's more complicated than I thought!

**Comment by tomwhite** [comment]
It might be simplest to just delegate to `dot` for the common 2D case, and use the existing implementation for other cases. I opened #6924 for this.

**Comment by ravwojdyla** [comment]
@tomwhite @mrocklin tried implementing `matmul` as `tensordot`, and it's possible but rather complicated and doesn't perform too well (for input above 2d, it get's complicated due to outer products nature of tensordot). So instead tried to implement a new matmul without concatenated contraction blockwise -> #7000.

**Comment by jsignell** [comment]
Was this closed by #7000?

**Comment by ravwojdyla** [comment]
@jsignell yes. And if I may add, I believe before the release, #7050 should be fixed/closed as well.
